# 05 — Régression Logistique sur les Rendements Bruts

## Objectif

Après avoir construit et évalué plusieurs stratégies de référence (baselines), l'objectif de ce notebook est d'étudier le premier modèle de Machine Learning du projet : la régression logistique.

Contrairement aux baselines précédentes, ce modèle apprend automatiquement une combinaison linéaire des rendements passés (`RET_1` à `RET_20`) afin d'estimer la probabilité conditionnelle qu'une allocation appartienne à la classe positive.

Mathématiquement, le modèle cherche à estimer :

$$
P(Y=1 \mid X)
$$

où \(X\) représente le vecteur des rendements historiques disponibles pour chaque allocation.

Cette expérimentation vise à déterminer si un modèle linéaire appris est capable d'exploiter davantage d'information contenue dans les rendements passés que les stratégies déterministes étudiées précédemment, tout en conservant un protocole d'évaluation strictement temporel afin d'éviter toute fuite d'information (*data leakage*).

---

## Question scientifique

Les rendements passés contiennent-ils suffisamment d'information pour qu'un modèle linéaire appris puisse estimer la probabilité d'une performance future positive de manière plus pertinente que des stratégies de décision déterministes ?

Plus précisément, ce notebook cherche à répondre aux questions suivantes :

- Les rendements passés (`RET_1` à `RET_20`) contiennent-ils un signal prédictif exploitable ?
- La régression logistique permet-elle d'améliorer les performances obtenues par les baselines précédemment étudiées ?
- Les probabilités estimées par le modèle permettent-elles d'établir un classement (*ranking*) des allocations plus informatif que les stratégies déterministes ?
- Le modèle conserve-t-il des performances stables lorsqu'il est évalué à l'aide d'une validation temporelle de type *expanding window* ?

---

## Protocole expérimental

Afin de garantir une comparaison équitable, reproductible et exempte de fuite d'information, le protocole expérimental suivant est appliqué tout au long de ce notebook :

- seules les variables de rendement brut (`RET_1` à `RET_20`) sont utilisées ;
- aucune étape de *feature engineering* n'est introduite à ce stade afin d'évaluer uniquement la capacité prédictive des variables d'origine ;
- la même stratégie de validation temporelle (*expanding window*) que celle utilisée pour les baselines est conservée ;
- toutes les étapes de prétraitement sont apprises exclusivement sur les données d'entraînement de chaque fold avant d'être appliquées au fold de validation ;
- le jeu de test officiel n'est jamais utilisé durant le développement du modèle ;
- toutes les performances présentées dans ce notebook sont calculées uniquement sur des données hors échantillon (*out-of-sample*).

## Architecture de modélisation

La logique réutilisable liée à la construction des modèles est placée dans `src/modeling.py`.

Ce choix évite de créer un fichier spécifique pour chaque algorithme et permet de conserver une architecture extensible pour les prochains modèles supervisés, comme Random Forest, Gradient Boosting ou XGBoost.

À ce stade, `modeling.py` contiendra uniquement les utilitaires nécessaires à la construction d'un pipeline de régression logistique, mais le fichier est conçu pour centraliser progressivement les composants liés à la modélisation.

In [1]:
import sys
from pathlib import Path

ROOT = Path.cwd().parent

if str(ROOT) not in sys.path:
    sys.path.append(str(ROOT))

In [2]:
import pandas as pd
import numpy as np
from src.data_loading import load_X_train, load_y_train
from src.modeling import build_logistic_regression_pipeline
from src.validation import create_expanding_window_folds, check_temporal_folds
from src.evaluation import evaluate_model_on_folds

In [3]:
X_train = load_X_train()
y_train = load_y_train()

In [4]:
df_train = X_train.merge(y_train,left_on="ROW_ID",right_on="ROW_ID")
df_train["class"] = np.where(df_train["target"] > 0,1,0)

In [5]:
df_train.columns

Index(['ROW_ID', 'TS', 'ALLOCATION', 'RET_20', 'RET_19', 'RET_18', 'RET_17',
       'RET_16', 'RET_15', 'RET_14', 'RET_13', 'RET_12', 'RET_11', 'RET_10',
       'RET_9', 'RET_8', 'RET_7', 'RET_6', 'RET_5', 'RET_4', 'RET_3', 'RET_2',
       'RET_1', 'SIGNED_VOLUME_20', 'SIGNED_VOLUME_19', 'SIGNED_VOLUME_18',
       'SIGNED_VOLUME_17', 'SIGNED_VOLUME_16', 'SIGNED_VOLUME_15',
       'SIGNED_VOLUME_14', 'SIGNED_VOLUME_13', 'SIGNED_VOLUME_12',
       'SIGNED_VOLUME_11', 'SIGNED_VOLUME_10', 'SIGNED_VOLUME_9',
       'SIGNED_VOLUME_8', 'SIGNED_VOLUME_7', 'SIGNED_VOLUME_6',
       'SIGNED_VOLUME_5', 'SIGNED_VOLUME_4', 'SIGNED_VOLUME_3',
       'SIGNED_VOLUME_2', 'SIGNED_VOLUME_1', 'MEDIAN_DAILY_TURNOVER', 'GROUP',
       'target', 'class'],
      dtype='str')

In [6]:
features_cols = ['RET_20', 'RET_19', 'RET_18', 'RET_17','RET_16', 
                 'RET_15', 'RET_14', 'RET_13', 'RET_12', 'RET_11', 
                 'RET_10','RET_9', 'RET_8', 'RET_7', 'RET_6', 
                 'RET_5', 'RET_4', 'RET_3', 'RET_2','RET_1'
                ]

In [7]:
dates = sorted(list(set(X_train["TS"])))
folds = create_expanding_window_folds(dates)
check_temporal_folds(folds,dates,validation_size=120)

True

In [8]:
folds 

[{'fold': 1,
  'train_start': 'DATE_0001',
  'train_end': 'DATE_2042',
  'valid_start': 'DATE_2043',
  'valid_end': 'DATE_2162'},
 {'fold': 2,
  'train_start': 'DATE_0001',
  'train_end': 'DATE_2162',
  'valid_start': 'DATE_2163',
  'valid_end': 'DATE_2282'},
 {'fold': 3,
  'train_start': 'DATE_0001',
  'train_end': 'DATE_2282',
  'valid_start': 'DATE_2283',
  'valid_end': 'DATE_2402'},
 {'fold': 4,
  'train_start': 'DATE_0001',
  'train_end': 'DATE_2402',
  'valid_start': 'DATE_2403',
  'valid_end': 'DATE_2522'}]

## Construction du pipeline

Le pipeline est composé de trois étapes successives :

- Imputation des valeurs manquantes (`SimpleImputer`)
- Standardisation des variables (`StandardScaler`)
- Classification (`LogisticRegression`)

L'ensemble des étapes de prétraitement est appris uniquement sur les données d'entraînement de chaque fold afin d'éviter toute fuite d'information.

In [9]:
logistic_regression_pipeline = build_logistic_regression_pipeline()

In [10]:
logistic_regression_results = evaluate_model_on_folds(df_train,folds,logistic_regression_pipeline,features_cols)

In [11]:
logistic_regression_results

,fold,train_start,train_end,valid_start,valid_end,train_positive_rate,valid_positive_rate,accuracy,log_loss,roc_auc,n_valid_predictions,n_correct_predictions
0,1,DATE_0001,DATE_2042,DATE_2043,DATE_2162,0.504803,0.522228,0.516588,0.692167,0.520935,24114,12457
1,2,DATE_0001,DATE_2162,DATE_2163,DATE_2282,0.505732,0.504837,0.529923,0.690788,0.540789,24396,12928
2,3,DATE_0001,DATE_2282,DATE_2283,DATE_2402,0.505687,0.520554,0.515708,0.692314,0.517971,24764,12771
3,4,DATE_0001,DATE_2402,DATE_2403,DATE_2522,0.506421,0.522097,0.519447,0.691120,0.528017,25660,13329


### Interprétation

La régression logistique obtient des performances relativement stables sur les quatre folds temporels.

Les résultats varient légèrement d'une période à l'autre, mais aucune amélioration systématique par rapport aux meilleures baselines n'apparaît.

Le modèle dépasse certaines baselines sur quelques folds et se fait dépasser sur d'autres, ce qui suggère que les rendements bruts contiennent un signal prédictif faible et peu stable dans le temps.

À ce stade, l'analyse fold par fold est principalement utile pour vérifier la stabilité temporelle du modèle plutôt que pour conclure sur la supériorité d'un modèle particulier.

In [12]:
logit_mean_accuracy = logistic_regression_results["accuracy"].mean()
print(f"La moyenne de l'accuracy de notre pipeline sur l'ensemble des folds est de {logit_mean_accuracy:.2f}")

La moyenne de l'accuracy de notre pipeline sur l'ensemble des folds est de 0.52


### Interprétation

La précision moyenne obtenue par la régression logistique est d'environ **52.04 %**.

Cette performance reste extrêmement proche de la meilleure baseline (Momentum : **52.13 %**).

L'utilisation simultanée des vingt rendements passés n'apporte donc pas d'amélioration significative par rapport à une règle de décision beaucoup plus simple reposant uniquement sur le signe de RET_1.

Ce résultat suggère que les rendements bruts, dans leur représentation actuelle, ne permettent pas au modèle linéaire d'extraire davantage d'information prédictive.

In [13]:
logit_std_accuracy = logistic_regression_results["accuracy"].std()
print(f"L'ecart-type de l'accuracy de notre pipeline sur l'ensemble des folds est de {logit_std_accuracy:.6f}")

L'ecart-type de l'accuracy de notre pipeline sur l'ensemble des folds est de 0.006536


### Interprétation

L'écart-type des performances entre les différents folds reste faible.

Cela indique que les performances de la régression logistique sont relativement stables au cours des différentes périodes temporelles étudiées.

Le problème principal ne semble donc pas être un manque de stabilité du modèle, mais plutôt une capacité prédictive globalement limitée.

In [14]:
logit_mean_roc_auc = logistic_regression_results["roc_auc"].mean()
logit_mean_log_loss = logistic_regression_results["log_loss"].mean()
pd.DataFrame({
    "accuracy" : logit_mean_accuracy,
    "roc_auc" : logit_mean_roc_auc,
    "log_loss" : logit_mean_log_loss
},index=["logistic_regression_pipeline"])


,accuracy,roc_auc,log_loss
logistic_regression_pipeline,0.520416,0.526928,0.691597


### Interprétation des métriques

Le ROC-AUC moyen est proche de 0.52.

Cela signifie que si l'on choisit aléatoirement une allocation positive et une allocation négative, le modèle attribuera un score plus élevé à l'allocation positive dans seulement environ 52 % des cas.

La capacité de discrimination reste donc très proche du hasard.

La Log-Loss reste proche de 0.693, valeur attendue pour un modèle attribuant une probabilité proche de 50 % à toutes les observations.

Les probabilités estimées par la régression logistique restent donc peu informatives.

# Conclusion

La régression logistique constitue le premier modèle supervisé étudié dans ce projet.

Malgré l'utilisation simultanée des vingt rendements passés (RET_1 à RET_20), le modèle n'obtient pas de gain significatif par rapport aux meilleures stratégies de référence.

Le ROC-AUC reste proche de 0.5 et la Log-Loss demeure proche de celle d'un modèle attribuant une probabilité proche de 50 % à chaque observation, ce qui suggère que les probabilités produites restent peu discriminantes.

Ces résultats ne permettent pas de conclure que les rendements passés ne contiennent aucun signal prédictif.

Ils suggèrent en revanche que les **rendements bruts**, utilisés directement comme variables explicatives, ne constituent probablement pas une représentation suffisamment informative pour le problème étudié.

La prochaine étape du projet consistera donc à concevoir de nouvelles variables (Feature Engineering) capables de mieux représenter la dynamique des marchés financiers avant d'envisager des modèles plus complexes.